offline/cable_input.F90

```diff
@@ -2887,5 +2921,402 @@ SUBROUTINE allocate_cable_vars(air,bgc,canopy,met,bal,                         &
+
+SUBROUTINE read_metvals(STD, DataArr, LandIDx, LandIDy, Year, DayOfYear,&
+    LeapYears)
+  ! Get the data from a day
+  TYPE(SPATIO_TEMPORAL_DATASET), INTENT(INOUT)  :: STD
+  REAL, DIMENSION(:), ALLOCATABLE, INTENT(INOUT)    :: DataArr
+  INTEGER, DIMENSION(:), POINTER, INTENT(IN)    :: LandIDx, LandIDy
+
+  INTEGER, INTENT(IN) :: DayOfYear, Year
+  LOGICAL, INTENT(IN) :: LeapYears
+
+  ! We'll need to compute the record index to grab
+  INTEGER     :: YearIndex, TimeIndex
+
+  ! Loop Iterators
+  INTEGER     :: FileIndx, VarNameIndx, YearIter, LandCell
+
+  ! Status checker
+  INTEGER  :: ok
+
+  TimeIndex = DayOfYear
+  YearIndex = Year
+
+  ! We've already opened a file, check whether the current year is open
+  IF ((Year >= STD%StartYear(STD%CurrentFileIndx)) .AND.&
+    (Year <= STD%EndYear(STD%CurrentFileIndx))) THEN
+    ! In this instance, we don't need to do anything
+    CONTINUE
+  ELSE
+    ok = NF90_CLOSE(STD%CurrentFileID)
+
+    ! First, check if we're outside the range of our data
+    IF (YEAR < STD%StartYear(1)) THEN
+      ! Before the first year
+      STD%CurrentFileIndx = 1
+      YearIndex = STD%StartYear(1)
+      TimeIndex = 1
+    ELSE IF (YEAR > STD%EndYear(SIZE(STD%EndYear))) THEN
+      ! After the last year
+      STD%CurrentFileIndx = SIZE(STD%EndYear)
+      YearIndex = STD%EndYear(STD%CurrentFileIndx)
+      IF ((LeapYears) .AND. (is_leapyear(YearIndex))) THEN
+        TimeIndex = 366
+      ELSE
+        TimeIndex = 365
+      END IF
+    ELSE
+      ! Normal operation, we're in the year of our data
+      FindFile: DO FileIndx = 1, SIZE(STD%FileNames)
+        IF ((Year >= STD%StartYear(FileIndx)) .AND. &
+          (Year <= STD%EndYear(FileIndx))) THEN
+          STD%CurrentFileIndx = FileIndx
+          EXIT FindFile
+        END IF
+      END DO FindFile
+    END IF
+
+    ! Open the new file
+    ok = NF90_OPEN(STD%FileNames(STD%CurrentFileIndx), NF90_NOWRITE,&
+      STD%CurrentFileID)
+
+    CALL find_variable_id(STD)
+  END IF
+
+  ! Now read the desired time step
+
+  ! Due to leapyears, we can't just do add 365 * number of years from startyear.
+  IF (LeapYears) THEN
+    CountDays: DO YearIter = STD%StartYear(STD%CurrentFileIndx), YearIndex
+      IF (is_leapyear(YearIndex)) THEN
+        TimeIndex = TimeIndex + 366
+      ELSE
+        TimeIndex = TimeIndex + 365
+      END IF
+    END DO CountDays
+  ELSE
+    TimeIndex = TimeIndex + 365 * (YearIndex - STD%StartYear(STD%CurrentFileIndx))
+  END IF
+  ! Now we have the index, we can grab the data
+
+  ! Read from the netCDF file to the masked array point by point
+  ApplyMask: DO LandCell = 1, SIZE(LandIDx)
+    ok = NF90_GET_VAR(STD%CurrentFileID,&
+      STD%CurrentVarID, DataArr(LandCell), START = (/LandIDx(LandCell),&
+      LandIDy(LandCell), TimeIndex/))
+      IF (ok /= NF90_NOERR) THEN
+        CALL handle_err(ok, "Failed reading "//STD%FileNames&
+          (STD%CurrentFileIndx)//" in read_metvals.")
+      END IF
+  END DO ApplyMask
+
```

Comment:
Is there the possibility to read the whole field at once, for example for a global run?
Reading point by point might be very costly if the input data is in netcdf4 and has a chunksize of the global field. Then you read SIZE(LandIDx) times a whole global field.

Reply:
Do you mean the functionality of the old DirectRead option? I didn't notice a big performance hit when I removed it, but I can add it back in.